README_tcpdevmem.md

```diff
@@ -0,0 +1,277 @@
+# Neper with TCPDirect run instructions
+
+Table of Contents
+- [Neper with TCPDirect run instructions](#neper-with-tcpdirect-run-instructions)
+  - [TCPDirect CUDA: tcp\_stream within Docker container](#tcpdirect-cuda-tcp_stream-within-docker-container)
+    - [Building your own image for testing](#building-your-own-image-for-testing)
+    - [Override CUDA library directory (DLVM)](#override-cuda-library-directory-dlvm)
+  - [TCPDirect UDMA: Compiling tcp\_stream](#tcpdirect-udma-compiling-tcp_stream)
+  - [Running tcp\_stream](#running-tcp_stream)
+    - [Added flags](#added-flags)
+    - [Running tcp\_stream via `multi_neper.py`](#running-tcp_stream-via-multi_neperpy)
+      - [Example of successful output](#example-of-successful-output)
+    - [Running tcp\_stream directly](#running-tcp_stream-directly)
+
+
+## TCPDirect CUDA: tcp_stream within Docker container
+
+A TCPDirect-Cuda enabled Docker image exists at `gcr.io/a3-tcpd-staging-hostpool/neper`. The `run_neper_container.sh` script makes it easy to run Neper:
+
+
+```
+./run_neper_container.sh
+
+# within the container
+FLOWS=2
```

Comment:
I think also a follow up work item, but run_neper_container.sh needs to know whether to run the release version of neper or a developer built version of neper.

the 'run_benchmark' script already does this. The stable version of the rxdm is at 'gcr.io/a3-tcpd-staging-hostpool/stable/tcpgpudmarxd'

and development versions are at something like:

'gcr.io/a3-tcpd-staging-hostpool/mina/tcpgpudmarxd'

We need a similar scheme for neper I think, but that is follow up work, I can file a bug.

Reply:
> if the run_neper_container.sh script can take all these as args

That would be if the container is created & removed for each tcp_stream run. The idea was that tcp_stream could be invoked multiple times within the same container

Reply:
Is there any damage from the container being created for each run?

FWIW each time rxdm or multinic benchmark is ran, the script is actually creating a container and running the binary in it, and it's all seemless to the user.

Having run_neper_container.sh (maybe renamed to run_neper.sh) take the args means that the user can run 1 command:

./run_neper_container.sh --devices dev --buffer-size....

Currently the user needs to run at the minimum 2 commands:

./run_neper_container.sh
./multi_neper.py ....

So by making run_neper_container.sh take args, we've reduced the number of manual steps needed to run neper by 50%.


The implementation is trivial btw, all that is needed is to add $@ to the end of the command that runs multi_neper.py. See run_benchmark or start_tcpdirectd_cos for an example on how to do it.

Reply:
I've modified run_neper_container.sh to take `$@` instead of hard-coding `bash`.

I've also added a note in the readme about creating a long-running container, which might be useful when access to files within the container is necessary (i.e. accessing log files when using `--tcpd-validate` flag).

Reply:
see https://github.com/google/neper/pull/46/commits/8b8a962921943ef2791e92edc39a071ee196078c